tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py

```diff
@@ -0,0 +1,95 @@
+import uuid
+from datetime import datetime
+from unittest import mock
+
+from django.conf import settings
+from django.test import override_settings
+
+from sentry.constants import DataCategory
+from sentry.ingest.billing_metrics_consumer import get_metrics_billing_consumer
+from sentry.sentry_metrics.indexer.strings import TRANSACTION_METRICS_NAMES
+from sentry.utils import json
+from sentry.utils.outcomes import Outcome
+
+
+@mock.patch("sentry.ingest.billing_metrics_consumer.track_outcome")
+def test_outcomes_consumed(track_outcome, kafka_producer, kafka_admin):
+    # Based on test_ingest_consumer_kafka.py
+
+    metrics_topic = "snuba-generic-metrics"
+
+    admin = kafka_admin(settings)
+    admin.delete_topic(metrics_topic)
+    producer = kafka_producer(settings)
+
+    with override_settings(
+        KAFKA_CONSUMER_AUTO_CREATE_TOPICS=True,
+    ):
+
+        buckets = [
+            {  # Counter metric with wrong ID will not generate an outcome
+                "metric_id": 123,
+                "type": "c",
+                "org_id": 1,
+                "project_id": 2,
+                "timestamp": 123,
+                "value": 123.4,
+            },
+            {  # Distribution metric with wrong ID will not generate an outcome
+                "metric_id": 123,
+                "type": "d",
+                "org_id": 1,
+                "project_id": 2,
+                "timestamp": 123,
+                "value": [1.0, 2.0],
+            },
+            {  # Empty distribution will not generate an outcome
+                # NOTE: Should not be emitted by Relay anyway
+                "metric_id": TRANSACTION_METRICS_NAMES["d:transactions/duration@millisecond"],
+                "type": "d",
+                "org_id": 1,
+                "project_id": 2,
+                "timestamp": 123,
+                "value": [],
+            },
+            {  # Valid distribution bucket emits an outcome
+                "metric_id": TRANSACTION_METRICS_NAMES["d:transactions/duration@millisecond"],
+                "type": "d",
+                "org_id": 1,
+                "project_id": 2,
+                "timestamp": 123456,
+                "value": [1.0, 2.0, 3.0],
+            },
+        ]
+
+        producer = kafka_producer(settings)
+        for bucket in buckets:
+            producer.produce(metrics_topic, json.dumps(bucket))
+        producer.flush()
+
+        metrics_consumer = get_metrics_billing_consumer(
+            topic=metrics_topic,
+            group_id=uuid.uuid4().hex,
+            auto_offset_reset="earliest",
+            force_topic=None,
+            force_cluster=None,
+        )
+
+        calls = track_outcome.mock_calls
+        for _ in range(100):
```

Comment:
I think this is going to be a fragile test as there is a waiting time every time you subscribe to a topic before you start receiving messages that depends on how long kafka takes to rebalance and assign you partitions. So 100 may be not enough in a flaky way.
Since you are mocking all the producer logic I would suggest you turn this into a test that does not depend on Kafka to be running.
Arroyo provides you three ways to test your consumer:
- unit test the processing strategy: This means instantiate `BillingTxCountMetricConsumerStrategy` with the factory you created and routinely call `submit` and `poll`. So you test your logic without the need of the entire Kafka scaffolding.
- test your consumer with the [arroyo in memory broker/consumer](https://github.com/getsentry/arroyo/blob/main/arroyo/backends/local/backend.py#L130). This is a fake of the Kafka consumer for unit tests so the response time is predictable.
- Or you test with the real Kafka but this should be used only when you want to run integration tests.

Reply:
Converted the test to a unit test calling `submit` and `poll` on the strategy repeatedly.